pkg/compactor/blocks_cleaner.go

```diff
@@ -100,31 +100,84 @@ func (c *BlocksCleaner) ticker(ctx context.Context) error {
 }
 
 func (c *BlocksCleaner) runCleanup(ctx context.Context) {
-	level.Info(c.logger).Log("msg", "started hard deletion of blocks marked for deletion")
+	level.Info(c.logger).Log("msg", "started hard deletion of blocks marked for deletion, and blocks for tenants marked for deletion")
 	c.runsStarted.Inc()
 
 	if err := c.cleanUsers(ctx); err == nil {
-		level.Info(c.logger).Log("msg", "successfully completed hard deletion of blocks marked for deletion")
+		level.Info(c.logger).Log("msg", "successfully completed hard deletion of blocks marked for deletion, and blocks for tenants marked for deletion")
 		c.runsCompleted.Inc()
 		c.runsLastSuccess.SetToCurrentTime()
 	} else if errors.Is(err, context.Canceled) {
-		level.Info(c.logger).Log("msg", "canceled hard deletion of blocks marked for deletion", "err", err)
+		level.Info(c.logger).Log("msg", "canceled hard deletion of blocks marked for deletion, and blocks for tenants marked for deletion", "err", err)
 		return
 	} else {
-		level.Error(c.logger).Log("msg", "failed to hard delete blocks marked for deletion", "err", err.Error())
+		level.Error(c.logger).Log("msg", "failed to hard delete blocks marked for deletion, and blocks for tenants marked for deletion", "err", err.Error())
 		c.runsFailed.Inc()
 	}
 }
 
 func (c *BlocksCleaner) cleanUsers(ctx context.Context) error {
-	users, err := c.usersScanner.ScanUsers(ctx)
+	users, deleted, err := c.usersScanner.ScanUsers(ctx)
 	if err != nil {
 		return errors.Wrap(err, "failed to discover users from bucket")
 	}
 
-	return concurrency.ForEachUser(ctx, users, c.cfg.CleanupConcurrency, func(ctx context.Context, userID string) error {
-		return errors.Wrapf(c.cleanUser(ctx, userID), "failed to delete user blocks (user: %s)", userID)
+	isDeleted := map[string]bool{}
+	for _, userID := range deleted {
+		isDeleted[userID] = true
+	}
+
+	allUsers := append(users, deleted...)
+	return concurrency.ForEachUser(ctx, allUsers, c.cfg.CleanupConcurrency, func(ctx context.Context, userID string) error {
+		if isDeleted[userID] {
+			return errors.Wrapf(c.cleanUserCompletely(ctx, userID), "failed to delete blocks for user marked for deletion: %s", userID)
+		}
+		return errors.Wrapf(c.cleanUser(ctx, userID), "failed to delete blocks for user: %s", userID)
+	})
+}
+
+// Remove all blocks for user marked for deletion.
+func (c *BlocksCleaner) cleanUserCompletely(ctx context.Context, userID string) error {
+	userLogger := util.WithUserID(userID, c.logger)
+	userBucket := bucket.NewUserBucketClient(userID, c.bucketClient)
+
+	level.Info(userLogger).Log("msg", "deleting blocks for user marked for deletion")
+
+	var deleted, failed int
+	err := userBucket.Iter(ctx, "", func(name string) error {
```

Comment:
Here we delete only blocks, but the users scanner will continue to report this user as far as there's an object in the user prefix (eg. markers/, debug/, bucket-index.json). How should that work?

_Not something we necessarily have to address in this PR, but we need a plan for that._

Reply:
My plan is this: when `deleteUser` is finished, it will update deletion mark with timestamp when it has finished.

When `deleteUser` is called next time, there are two cases:
- if there are more blocks to delete, `deleteUser` will delete them, and update tenant deletion mark again with "finished" time.
- if no more blocks are to be deleted, `deleteUser` will check "finished" time in the marker, and after some time (say a week), it will do remaining cleanup work: deleting objects with the "user/" prefix completely (with marker file being the last one).

Reply:
> it will update deletion mark with timestamp when it has finished

Uhm. This makes the deletion mark no more immutable. Let's talk about it offline. Not to be addressed in this PR.

Reply:
That's correct. I don't think it's a problem -- other components just check for existence. Compactor will be the only component actually reading the tenant deletion mark.

If you think this is still an issue, we could also write "finished" timestamp into a separate file.

Reply:
> Compactor will be the only component actually reading the tenant deletion mark.

True. You're right, probably not an issue. I was thinking about adding it to the bucket index too, but we actually just need to store the "existence" in the index.